### PR TITLE
tail: fix race condition

### LINKED ISSUE #3765: GNU tests/tail-2/F-headers is failing often in the GNU CI
`Error: GNU test failed: tests/tail-2/F-headers. tests/tail-2/F-headers is passing on 'main'. Maybe you have to rebase?`

The error is:

```
2022-08-01T07:32:10.8281270Z FAIL: tests/tail-2/F-headers
2022-08-01T07:32:10.8281507Z ============================
2022-08-01T07:32:10.8281643Z 
2022-08-01T07:32:10.8281881Z tail: cannot open 'b' for reading: No such file or directory
2022-08-01T07:32:10.8282294Z ./tests/tail-2/F-headers.sh: a: unexpected delay?
2022-08-01T07:32:10.8282700Z tail: cannot open 'a' for reading: No such file or directory
2022-08-01T07:32:10.8283128Z tail: cannot open 'b' for reading: No such file or directory
2022-08-01T07:32:10.8283395Z ==> b <==
2022-08-01T07:32:10.8283713Z FAIL tests/tail-2/F-headers.sh (exit status: 1)
2022-08-01T07:32:10.8283900Z 
```

@niyaznigmatullin tried to fix it in https://github.com/uutils/coreutils/pull/3737

### PR DESCRIPTION
**Pull request by jhscheer**
fixes #3765

There exists a race condition (RC) that can occur if changes to a path
happen after the initial print loop in `uu_tail()`, but before the
path is added to the notify-Watcher thread in `follow()`.

@niyaznigmatullin found this RC and [identified](https://github.com/uutils/coreutils/issues/3765#issuecomment-1206154992) it as the reason why "gnu/tests/tail-2/F-headers.sh" failed sometimes in the CI and he also came up with a [fix](https://github.com/uutils/coreutils/pull/3786) for it.

The proposed fix manually handles possible events that could have been missed before.
However, this does not reduce the possibility for the RC to occur, and it uses a lot of duplicate code instead of utilizing the functionality of the Watcher thread together with the `handle_event()` function.

To minimize the window where the RC can occur, this PR moves starting the
Watcher thread and adding paths to it from `follow()` to the initial
print loop in `uu_tail()`.

Additionally, to make sure the RC cannot happen in
"gnu/tests/tail-2/F-headers.sh", the error message that is used as a trigger
in this test, is delayed until the path is added to the Watcher thread.

The workarounds for "tests/tail-2/F-headers.sh" in build-gnu.sh should no longer be needed.

### CODE REVIEW

**Comment by sylvestre** [comment]
@niyaznigmatullin how do you feel about this? thanks

File: src/uu/tail/src/tail.rs [hunk 1]
```diff
@@ -458,6 +468,9 @@ fn uu_tail(mut settings: Settings) -> UResult<()> {
                         unbounded_tail(&mut reader, &settings)?;
                     }
                     if settings.follow.is_some() {
+                        if let Some((ref mut watcher, _)) = watcher_rx {
```
Comment (niyaznigmatullin) [inline src/uu/tail/src/tail.rs:471]: Can we move this one before tail is actually printed, so that we don't have anytime that event can happen in between?

**Comment by niyaznigmatullin** [comment]
I like the change, but can we move the watcher addition before tail actually writes anything?

**Comment by niyaznigmatullin** [comment]
I think I found another race condition on detecting the orphans. Because it's detected after follow is run, if a file didn't exist and it's parent directory didn't exist, we print that we cannot access it. And the if it's created, we don't watch it, and it's not detecting that it's an orphan, because the directory could have been created by that time. 

As I understand, the idea was, that we need to make sure for each file either it's being watched, or it's parent is watched or it's added to orphans. So the last is not true sometimes.

**Comment by jhscheer** [comment]
@niyaznigmatullin Thanks for the feedback. I think both of your points are valid. I'll implement them soon.

**Comment by jhscheer** [comment]
This should be good to go.

**Comment by niyaznigmatullin** [comment]
I tested all tests with timeouts, everything passed but `descriptor-vs-rename.sh`. I've found the reason, it moves the file:
`mv a b` and then immediately writes `echo y >> b`, but when we handle event of renaming, we seek to the end of the new file, skipping the last written part, which never occurs in tail output.

I propose to use old opened reader, since it stores the file descriptor inside it, so no need to recreate one. Proposed change is here: https://github.com/niyaznigmatullin/coreutils/commit/fd145ee1e53e7fc2a6bce9a8c56563bf7bf9f8cb

**Comment by jhscheer** [comment]
> I tested all tests with timeouts, everything passed

Nice, thanks for testing.

> I propose to use old opened reader, since it stores the file descriptor inside it, so no need to recreate one. Proposed change is here: [niyaznigmatullin@fd145ee](https://github.com/niyaznigmatullin/coreutils/commit/fd145ee1e53e7fc2a6bce9a8c56563bf7bf9f8cb)

That looks like a good idea, please open a follow-up PR.
(I would keep the `?` for the `File::open()` call and remove the `None` case. Otherwise if there is an error, we will never know about it.)

**Pull request merged**